### DISCUSSION TITLE
Rationale behind using LibreWolf's portable launcher

### DISCUSSION THREAD

**Discussion by ltguillaume** [ScoopInstaller/Scoop#5622] [General]
Using LibreWolf via scoop raises the following issues:

1. People who pin the browser to the taskbar end up using a new profile being created in %AppData%, instead of the one in `persist\librewolf`, _or_ they end up with two icons on the taskbar when LibreWolf is running (the launcher's and the browser's).
2. Registering file/protocol associations from within the browser causes the profile in %AppData% to be used.
3. All kinds of semi-working solutions are offered instead (see https://github.com/ScoopInstaller/Extras/issues/7873).

Source: https://old.reddit.com/r/LibreWolf/comments/162vam1/librewolf_fails_to_update_properly_when_using/

These issues are caused by using the portable launcher I created, instead of `librewolf.exe` directly. The portable launcher is primarily for use on portable drives and for people who don't want _any_ system integration to happen (so file/protocol association, pinning to the taskbar etc. naturally are all out of its scope). Scoop, on the other hand, is only "semi-portable": it doesn't use installers if it doesn't have to and tries to place all critical application data into a single folder (`scoop\persist`), but apart from that, it adds Start menu shortcuts, adds shims to run the applications from anywhere by their executable's name and allows the apps to interact/integrate with the OS freely.

Then why not just use the portable archive to extract the browser, but use `librewolf.exe` instead of the launcher and make sure the profile is created in `persist\librewolf` the same way it's done for Firefox?

**Comment by HUMORCE** [comment]
https://github.com/ScoopInstaller/Extras/issues/8779#issuecomment-1171377205

**Reply by ltguillaume** [reply]
>>For pre_install, we could probably open $Env:AppData\Mozilla\Firefox\profiles.ini, and set the "Scoop profile" as default. (put Default=1 under it), but I am not sure if there is any potential problem if we do this.
>
>firefox cannot share profiles between individual installations.
>
><SCOOP>\apps\firefox\100.0
><SCOOP>\apps\firefox\102.0
>
>each firefox update via scoop is done as a individual installation, so profiles.ini will be ignored. even if this method were possible, this would change the last profile(default) selected by users.

This wouldn't apply to LibreWolf if you stick to using the portable archive for extracting/updating, so it could still be a solution. It is pretty much exactly what I do for LibreWolf on one system.

**Reply by HUMORCE** [reply]
Almost all browser manifests facing these problem. for primary browser, thier official installer is recommend.

Firefox-based browsers is more suitable install through Scoop, if the user understands how to manage Profiles. instead, all chromium-based browsers: https://github.com/ScoopInstaller/Extras/issues/5742

Is there a way to lock or set Firefox's profile for a individual installation?

---

`extras/librewolf` is not handled in same way as `extras/firefox`, simply becuase upstream privided the Portable Launcher you created.

**Reply by ltguillaume** [reply]
Crap, you're right. I believe that just changing the `profiles.ini` beforehand used to work just fine, but somewhere along the line they introduced the unique installation identifiers and now it's hard to do anything like that beforehand.

So, what I see in the json for Firefox is that while you create a profile in `persist\firefox`, if the user just starts Firefox the normal way (without first opening the profile manager), that profile is never used. Instead, a new profile is created in %AppData% upon first run.

> Is there a way to lock or set Firefox's profile for a individual installation?

While not ideal, I think a good workaround would to be to run `firefox.exe -p Scoop` right after. The other default profile in %AppData% would then not be created at all.

**Reply by ltguillaume** [reply]
> `extras/librewolf` is not handled in same way as `extras/firefox`, simply becuase upstream privided the Portable Launcher you created.

Yeah I know. Not many people must use the `librewolf.exe` shim by the way, because when I tested the portable launcher through that, I saw that it introduced an issue with it (the shim waiting for the launcher and the launcher waiting for the shim to close). I fixed it in https://codeberg.org/ltguillaume/librewolf-portable/releases/tag/1.6.5 just to be sure, but as said, I'm not sure if using the portable launcher for "semi-portable" use cases is all that great.

**Reply by ltguillaume** [reply]
@HUMORCE I've tested it using https://github.com/ltguillaume/schep/blob/main/bucket/firefox-esr.json for Firefox and https://github.com/ltguillaume/schep/blob/main/bucket/betterbird-future.json for Betterbird (=Thunderbird) and this works nicely: profile is created and made default by running the program with `-P Scoop` (or `-P Scoop-ESR`).

**Reply by HUMORCE** [reply]
No, it does not work as you described. It may be that you have already set default profile to `Scoop-ESR` for the individual installation(`...\firefox-esr\115.2.0`), before you test.

In a fresh environment:
1. scoop install https://github.com/ltguillaume/schep/blob/main/bucket/firefox-esr.json
2. firefox-esr runs with profile `Scoop-ESR` in post_install stage.
3. Close app
4. Open firefox-esr via shortcut or command
5. firefox-esr runs with profile `default-esr`

**Reply by ltguillaume** [reply]
No, this was all tested on a clean VM without scoop or Firefox/LibreWolf even installed yet.

I did indeed follow the exact steps as you described for this.

The only reason I can think of it did not work as expected for you, is that there was already an `%AppData%\Mozilla\Firefox\profiles.ini` file present.

Test environment:
Windows 10 LTSC2021 on VirtualBox
Scoop installed via default method `irm get.scoop.sh | iex`
Tried using PowerShell and command line prompt FWIW
Worked as expected for Firefox-ESR (EN/NL), LibreWolf and Betterbird (EN/NL) (all in new environments, i.e. without another FF/LW/BB instance already installed, either via Scoop or separately).

__Additional question__: why is there a symlink from `scoop\apps\firefox-esr\profile` to `scoop\persist\firefox-esr\profile`? This is never used by Firefox itself.

**Reply by HUMORCE** [reply]
> The only reason I can think of it did not work as expected for you, is that there was already an %AppData%\Mozilla\Firefox\profiles.ini file present.

Yes, but we cannot delete this file in each installation.

> why is there a symlink from scoop\apps\firefox-esr\profile to scoop\persist\firefox-esr\profile? This is never used by Firefox itself.

Long time ago, the path of portable profile was set to `$dir\profile` instead of `$persist_dir\profile`, it created for easy access by users.

**Reply by ltguillaume** [reply]
> > The only reason I can think of it did not work as expected for you, is that there was already an %AppData%\Mozilla\Firefox\profiles.ini file present.
> 
> Yes, but we cannot delete this file in each installation.

I understand that, but it's not really the point of this change. The goal is to have the Scoop(-ESR) profile set as default on a clean installation, so that

1. All the people starting to use Firefox won't experience the issue that their actual profile is in %AppData%, while an unused profile is created in `$persist_dir\profile`. This not only leads to confusion, but also to wrongfully thinking their profile is safe by backing up `scoop\persist`, while in fact the profile they're using isn't in there.
2. All the people using LibreWolf don't have to use the portable launcher, which in the use case of scoop is only annoying.

> > why is there a symlink from scoop\apps\firefox-esr\profile to scoop\persist\firefox-esr\profile? This is never used by Firefox itself.
> 
> Long time ago, the path of portable profile was set to `$dir\profile` instead of `$persist_dir\profile`, it created for easy access by users.

Ah, fair enough. I thought I'd remove it then, but I see that Firefox doesn't create the profile without the folder `$persist_dir\profile` already existing. So it's either the `persist` entry or an explicit `mkdir` that is still required.

**Reply by HUMORCE** [reply]
> I understand that, but it's not really the point of this change. The goal is to have the Scoop(-ESR) profile set as default on a clean installation, so that

Then the manifest get different behavior between different users.

**Reply by ltguillaume** [reply]
True, just like the current one already does, depending on whether they manually open the profile manager beforehand (a requirement) or not. At least this way, new users won't have to do anything.

**Comment by JustArion** [comment]
I'm not really understanding the discussion so far. Has the problem been fixed? My current problems I have with the scoop install is what @ltguillaume noted in the start of the discussion. 

- I have to re-pin my current version on the Taskbar to prevent 2 Taskbar icons
- I have to change the profile being used every update back to the one I'm using

Were these issues addressed over the years? If they were, how can I remedy it?

**Reply by ltguillaume** [reply]
No. I just started using my own bucket for LibreWolf, Firefox-ESR, and Betterbird, which only partly circumvents the issue: it needs to start the profile manager after an update, so that the user can click OK to make sure the Scoop profile is used, otherwise a new profile is created.

```
scoop bucket add schep https://github.com/ltguillaume/schep
scoop install schep/librewolf
```

**Reply by HUMORCE** [reply]
> Were these issues addressed over the years? If they were, how can I remedy it?

Best Practice: 

Do not install main browser via Scoop. Almost all browsers don't benefit from Scoop's features.

The default profile of Firefox-based browser can be specified with profile selector, but Chromium-based browsers require more complex modification (shell registries, shortcut's launch option, cli flags, etc.)

If you're just afraid of losing your Firefox Profile, you can always create a profile in a standard installation of FireFox that points to `<SCOOP>\persists\firefox\profile\<profile_name>`.

**Reply by ltguillaume** [reply]
>Do not install main browser via Scoop. Almost all browsers don't benefit from Scoop's features.

@HUMORCE Well, LibreWolf is a browser that doesn't have an internal updater, so Scoop could be pretty useful if it weren't using directory junctions for the `current` folder (or if Mozilla applications didn't respond to that by assuming a fresh installation because of it).

For Firefox, I would also advise most users to just use its internal updater or its bundled [Maintenance Service](https://support.mozilla.org/en-US/kb/what-mozilla-maintenance-service).
